jax/interpreters/xla.py

```diff
@@ -533,6 +602,52 @@ def _xla_callable(fun: lu.WrappedFun, device, backend, name, *arg_specs):
   else:
     return partial(_execute_replicated, compiled, result_handlers)
 
+def configure_aliasing(
+    c: xc.XlaBuilder,
+    xla_args: Sequence[xe.XlaOp],
+    out_tuple: xe.XlaOp,
+    arg_donate: Sequence[bool],
+    tuple_args: bool,
+):
+  """Configures input/output "must" aliasing based on `arg_donate`."""
+  assert len(xla_args) == len(arg_donate)
+
+  # First for every input array add it to `donations` iff it is a member of
+  # `arg_donate`.
+  donations = defaultdict(deque)
+  for arg_index, arg in enumerate(xla_args):
+    if arg_donate[arg_index]:
+      for param_index, element in flatten_shape(c.GetShape(arg)):
+        key = (element.dimensions(), element.numpy_dtype())
+        if tuple_args:
+          param_number = 0
+          param_index = (arg_index,) + tuple(param_index)
+          donations[key].append((param_number, param_index, arg_index))
+        else:
+          param_number = arg_index
+          donations[key].append((param_number, param_index, arg_index))
+
+  # Consume donations for outputs.
+  for output_index, element in flatten_shape(c.GetShape(out_tuple)):
+    key = (element.dimensions(), element.numpy_dtype())
+    if donations.get(key, ()):
+      param_number, param_index, _ = donations[key].popleft()
+      c.SetUpAlias(output_index, param_number, param_index)
+
+  # Check that all donations have been consumed such that buffer donation
+  # "must alias".
+  leftovers = tree_leaves([tuple(v) for v in donations.values()])
+  if leftovers:
+    leftovers = "\n".join([
+        f"- flat_args[{arg_index}] (spec: {dtype.name}{list(dims)})"
+        for (dims, dtype), donation in donations.items()
+        for _, _, arg_index in donation])
+
+    raise ValueError(
```

Comment:
The phrasing of the API as "donation" makes this somewhat surprising. I'm presumably free to throw your donation away, in the usual meaning of those words?

Reply:
Per the design review I think we're good with "donation" being "use if you like" (allowing us to not use them at all on CPU/GPU today and still be "in contract")/